File: 59. Spiral Matrix II/main.py
from typing import List


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        result = [[0] * n for _ in range(n)]
        left = 0
        right = n - 1
        top = 0
        bottom = n - 1
        num = 1

        while left <= right and top <= bottom:
            for i in range(left, right + 1):
                result[top][i] = num
                num += 1
            top += 1

            for i in range(top, bottom + 1):
                result[i][right] = num
                num += 1
            right -= 1

            if top <= bottom:
                for i in range(right, left - 1, -1):
                    result[bottom][i] = num
                    num += 1
                bottom -= 1

            if left <= right:
                for i in range(bottom, top - 1, -1):
                    result[i][left] = num
                    num += 1
                left += 1

        return result


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        result = [[0] * n for _ in range(n)]
        left = 0
        right = n - 1
        top = 0
        bottom = n - 1
        num = 1
        total = n * n

        while num <= total:
            for i in range(left, right + 1):
                result[top][i] = num
                num += 1
            top += 1

            for i in range(top, bottom + 1):
                result[i][right] = num
                num += 1
            right -= 1

            for i in range(right, left - 1, -1):
                result[bottom][i] = num
                num += 1
            bottom -= 1

            for i in range(bottom, top - 1, -1):
                result[i][left] = num
                num += 1
            left += 1

        return result


class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        result = [[0] * n for _ in range(n)]
        left = 0
        right = n - 1
        top = 0
        bottom = n - 1
        num = 1
        total = n * n
        direction = 1

        while num <= total:
            if direction == 1:
                for i in range(left, right + 1):
                    result[top][i] = num
                    num += 1
                top += 1
                direction = 2
            elif direction == 2:
                for i in range(top, bottom + 1):
                    result[i][right] = num
                    num += 1
                right -= 1
                direction = 3
            elif direction == 3:
                for i in range(right, left - 1, -1):
                    result[bottom][i] = num
                    num += 1
                bottom -= 1
                direction = 4
            else:
                for i in range(bottom, top - 1, -1):
                    result[i][left] = num
                    num += 1
                left += 1
                direction = 1

        return result

File: 59. Spiral Matrix II/main.rs
struct Solution;

impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut result = vec![vec![0; n as usize]; n as usize];
        let mut left = 0;
        let mut right = n - 1;
        let mut top = 0;
        let mut bottom = n - 1;
        let mut num = 1;

        while left <= right && top <= bottom {
            for i in left..=right {
                result[top as usize][i as usize] = num;
                num += 1;
            }
            top += 1;

            for i in top..=bottom {
                result[i as usize][right as usize] = num;
                num += 1;
            }
            right -= 1;

            if top <= bottom {
                for i in (left..=right).rev() {
                    result[bottom as usize][i as usize] = num;
                    num += 1;
                }
                bottom -= 1;
            }

            if left <= right {
                for i in (top..=bottom).rev() {
                    result[i as usize][left as usize] = num;
                    num += 1;
                }
                left += 1;
            }
        }

        result
    }
}

impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut result = vec![vec![0; n as usize]; n as usize];
        let mut left = 0;
        let mut right = n - 1;
        let mut top = 0;
        let mut bottom = n - 1;
        let mut num = 1;
        let total = n * n;

        while num <= total {
            for i in left..=right {
                result[top as usize][i as usize] = num;
                num += 1;
            }
            top += 1;

            for i in top..=bottom {
                result[i as usize][right as usize] = num;
                num += 1;
            }
            right -= 1;

            for i in (left..=right).rev() {
                result[bottom as usize][i as usize] = num;
                num += 1;
            }
            bottom -= 1;

            for i in (top..=bottom).rev() {
                result[i as usize][left as usize] = num;
                num += 1;
            }
            left += 1;
        }

        result
    }
}

impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut result = vec![vec![0; n as usize]; n as usize];
        let mut left = 0;
        let mut right = n - 1;
        let mut top = 0;
        let mut bottom = n - 1;
        let mut num = 1;
        let total = n * n;
        let mut direction = 1;

        while num <= total {
            match direction {
                1 => {
                    for i in left..=right {
                        result[top as usize][i as usize] = num;
                        num += 1;
                    }
                    top += 1;
                    direction = 2;
                }
                2 => {
                    for i in top..=bottom {
                        result[i as usize][right as usize] = num;
                        num += 1;
                    }
                    right -= 1;
                    direction = 3;
                }
                3 => {
                    for i in (left..=right).rev() {
                        result[bottom as usize][i as usize] = num;
                        num += 1;
                    }
                    bottom -= 1;
                    direction = 4;
                }
                _ => {
                    for i in (top..=bottom).rev() {
                        result[i as usize][left as usize] = num;
                        num += 1;
                    }
                    left += 1;
                    direction = 1;
                }
            }
        }

        result
    }
}
